File: tests/ui-testing/playwright-tests/utils/sourcemap-helpers.js
/**
 * Sourcemap Test Helpers
 *
 * Shared utility functions for sourcemap E2E testing.
 * Handles test app building, sourcemap upload/delete, stacktrace resolution, and user management.
 */

const fs = require('fs').promises;
const { exec } = require('child_process');
const { promisify } = require('util');
const path = require('path');
const execAsync = promisify(exec);

const testLogger = require('./test-logger');

/**
 * Build test app with updated RUM SDK configuration
 * @param {string} rumToken - RUM token from API
 * @param {Object} options - Build options
 * @returns {Promise<Object>} Build result with hashes and paths
 */
async function buildTestApp(rumToken, options = {}) {
  const {
    site = 'localhost:5080',
    insecureHTTP = true,
    service = 'o2-sourcemap-test-app',
    env = 'testing',
    version = '1.0.0-automation',
    org = 'default'
  } = options;

  testLogger.info('Building test app with RUM SDK configuration');

  const appDir = path.join(__dirname, '../../../MD_Files/Sourcemaps/o2-sourcemap-test-app');
  const configPath = path.join(appDir, 'src/index.js');

  try {
    // Read current config
    let configContent = await fs.readFile(configPath, 'utf8');

    // Update SDK config
    configContent = configContent.replace(
      /clientToken:\s*['"][^'"]*['"]/,
      `clientToken: '${rumToken}'`
    );
    configContent = configContent.replace(
      /site:\s*['"][^'"]*['"]/,
      `site: '${site}'`
    );
    configContent = configContent.replace(
      /insecureHTTP:\s*(?:true|false)/,
      `insecureHTTP: ${insecureHTTP}`
    );
    configContent = configContent.replace(
      /service:\s*['"][^'"]*['"]/,
      `service: '${service}'`
    );
    configContent = configContent.replace(
      /env:\s*['"][^'"]*['"]/,
      `env: '${env}'`
    );
    configContent = configContent.replace(
      /version:\s*['"][^'"]*['"]/,
      `version: '${version}'`
    );
    configContent = configContent.replace(
      /organizationIdentifier:\s*['"][^'"]*['"]/,
      `organizationIdentifier: '${org}'`
    );

    // Write updated config
    await fs.writeFile(configPath, configContent, 'utf8');
    testLogger.info('Updated SDK configuration');

    // Build app
    testLogger.info('Running npm run build...');
    const { stdout, stderr } = await execAsync('npm run build', {
      cwd: appDir,
      timeout: 60000 // 60 second timeout
    });

    if (stderr && !stderr.includes('webpack')) {
      testLogger.warn(`Build warnings: ${stderr}`);
    }

    testLogger.info('Build completed successfully');

    // Extract content hashes from dist/
    const distPath = path.join(appDir, 'dist');
    const distFiles = await fs.readdir(distPath);

    const hashes = {
      main: extractHash(distFiles, 'main', '.js'),
      lazyModule: extractHash(distFiles, 'lazy-module', '.js'),
      profiler: extractHash(distFiles, 'profiler', '.js'),
      recorder: extractHash(distFiles, 'recorder', '.js')
    };

    testLogger.info(`Extracted hashes: ${JSON.stringify(hashes)}`);

    // Create sourcemaps ZIP
    const sourcemapsZipPath = path.join(distPath, 'sourcemaps.zip');

    // Remove old ZIP if exists
    try {
      await fs.unlink(sourcemapsZipPath);
    } catch (err) {
      // Ignore if doesn't exist
    }

    await execAsync('zip sourcemaps.zip *.map', { cwd: distPath });
    testLogger.info(`Created sourcemaps ZIP at ${sourcemapsZipPath}`);

    return {
      appDir,
      distPath,
      hashes,
      sourcemapsZipPath,
      config: { service, env, version, org }
    };

  } catch (error) {
    testLogger.error(`Failed to build test app: ${error.message}`);
    throw error;
  }
}

/**
 * Extract content hash from filename
 * @param {string[]} files - List of filenames
 * @param {string} prefix - File prefix (e.g., 'main')
 * @param {string} ext - File extension (e.g., '.js')
 * @returns {string} Extracted hash
 */
function extractHash(files, prefix, ext) {
  const pattern = new RegExp(`${prefix}\\.(.*?)\\${ext}`);
  const file = files.find(f => f.startsWith(prefix) && f.endsWith(ext) && !f.endsWith('.map'));

  if (!file) {
    throw new Error(`Could not find file matching ${prefix}*${ext}`);
  }

  const match = file.match(pattern);
  if (!match || !match[1]) {
    throw new Error(`Could not extract hash from ${file}`);
  }

  return match[1];
}

/**
 * Serve test app on specified port
 * @param {string} distPath - Path to dist directory
 * @param {number} port - Port number (default: 8089)
 * @returns {Promise<Object>} Server process info
 */
async function serveTestApp(distPath, port = 8089) {
  testLogger.info(`Starting HTTP server on port ${port}...`);

  const { spawn } = require('child_process');

  const serverProcess = spawn('python3', ['-m', 'http.server', port.toString()], {
    cwd: distPath,
    detached: true,
    stdio: 'ignore'
  });

  serverProcess.unref();

  // Wait for server to be ready
  await new Promise(resolve => setTimeout(resolve, 2000));

  testLogger.info(`HTTP server started (PID: ${serverProcess.pid})`);

  return {
    process: serverProcess,
    pid: serverProcess.pid,
    port,
    url: `http://localhost:${port}`
  };
}

/**
 * Stop test app HTTP server
 * @param {Object} serverInfo - Server info from serveTestApp
 */
async function stopTestAppServer(serverInfo) {
  if (!serverInfo || !serverInfo.pid) {
    testLogger.warn('No server PID provided, skipping stop');
    return;
  }

  testLogger.info(`Stopping HTTP server (PID: ${serverInfo.pid})...`);

  try {
    // Kill process group
    process.kill(-serverInfo.pid, 'SIGTERM');
    testLogger.info('HTTP server stopped');
  } catch (error) {
    testLogger.warn(`Failed to stop server: ${error.message}`);
    // Try direct kill
    try {
      process.kill(serverInfo.pid, 'SIGKILL');
    } catch (err) {
      testLogger.error(`Could not kill server process: ${err.message}`);
    }
  }

  // Also try to kill any python processes on the port
  try {
    await execAsync(`lsof -ti:${serverInfo.port} | xargs kill -9 2>/dev/null || true`);
  } catch (err) {
    // Ignore errors
  }
}

/**
 * Upload sourcemaps to OpenObserve
 * @param {Object} request - Playwright APIRequestContext
 * @param {Object} config - Upload configuration
 * @param {string} zipPath - Path to sourcemaps ZIP
 * @returns {Promise<Object>} Response
 */
async function uploadSourcemaps(request, config, zipPath) {
  const { baseUrl, org, auth, service, env, version } = config;

  testLogger.info(`Uploading sourcemaps: service=${service}, env=${env}, version=${version}`);

  const FormData = require('form-data');
  const form = new FormData();

  // Read ZIP file
  const zipBuffer = await fs.readFile(zipPath);
  form.append('file', zipBuffer, {
    filename: 'sourcemaps.zip',
    contentType: 'application/zip'
  });
  form.append('service', service);
  form.append('env', env);
  form.append('version', version);

  const response = await request.post(`${baseUrl}/api/${org}/sourcemaps`, {
    headers: {
      'Authorization': `Basic ${auth}`,
      ...form.getHeaders()
    },
    data: form.getBuffer()
  });

  const status = response.status();
  testLogger.info(`Upload response: ${status}`);

  if (status !== 201 && status !== 200) {
    const body = await response.text();
    testLogger.error(`Upload failed: ${body}`);
    throw new Error(`Upload failed with status ${status}: ${body}`);
  }

  return response;
}

/**
 * List sourcemaps from OpenObserve
 * @param {Object} request - Playwright APIRequestContext
 * @param {Object} config - List configuration
 * @returns {Promise<Array>} List of sourcemaps
 */
async function listSourcemaps(request, config) {
  const { baseUrl, org, auth, service, env, version } = config;

  const params = new URLSearchParams();
  if (service) params.append('service', service);
  if (env) params.append('env', env);
  if (version) params.append('version', version);

  const url = `${baseUrl}/api/${org}/sourcemaps?${params}`;
  testLogger.info(`Listing sourcemaps: ${url}`);

  const response = await request.get(url, {
    headers: { 'Authorization': `Basic ${auth}` }
  });

  if (response.status() !== 200) {
    const body = await response.text();
    testLogger.error(`List failed: ${body}`);
    throw new Error(`List failed with status ${response.status()}`);
  }

  return await response.json();
}

/**
 * Delete sourcemaps from OpenObserve
 * @param {Object} request - Playwright APIRequestContext
 * @param {Object} config - Delete configuration
 * @returns {Promise<Object>} Response
 */
async function deleteSourcemaps(request, config) {
  const { baseUrl, org, auth, service, env, version } = config;

  const params = new URLSearchParams({ service, env, version });
  const url = `${baseUrl}/api/${org}/sourcemaps?${params}`;

  testLogger.info(`Deleting sourcemaps: ${url}`);

  const response = await request.delete(url, {
    headers: { 'Authorization': `Basic ${auth}` }
  });

  const status = response.status();
  testLogger.info(`Delete response: ${status}`);

  if (status !== 200) {
    const body = await response.text();
    testLogger.warn(`Delete returned non-200: ${body}`);
  }

  return response;
}

/**
 * Resolve stacktrace via sourcemap API
 * @param {Object} request - Playwright APIRequestContext
 * @param {Object} config - Resolution configuration
 * @param {string} stacktrace - Minified stacktrace
 * @returns {Promise<Object>} Resolved stacktrace
 */
async function resolveStacktrace(request, config, stacktrace) {
  const { baseUrl, org, auth, service, env, version } = config;

  testLogger.info(`Resolving stacktrace for ${service}:${version}`);

  const response = await request.post(`${baseUrl}/api/${org}/sourcemaps/stacktrace`, {
    headers: {
      'Authorization': `Basic ${auth}`,
      'Content-Type': 'application/json'
    },
    data: {
      service,
      env,
      version,
      stacktrace
    }
  });

  if (response.status() !== 200) {
    const body = await response.text();
    testLogger.error(`Resolution failed: ${body}`);
    throw new Error(`Resolution failed with status ${response.status()}`);
  }

  return await response.json();
}

/**
 * Get RUM token from OpenObserve
 * @param {Object} request - Playwright APIRequestContext
 * @param {Object} config - Configuration
 * @returns {Promise<string>} RUM token
 */
async function getRumToken(request, config) {
  const { baseUrl, org, auth } = config;

  testLogger.info('Fetching RUM token...');

  const response = await request.get(`${baseUrl}/api/${org}/rumtoken`, {
    headers: { 'Authorization': `Basic ${auth}` }
  });

  if (response.status() !== 200) {
    const body = await response.text();
    testLogger.error(`Failed to get RUM token: ${body}`);
    throw new Error(`Failed to get RUM token: ${response.status()}`);
  }

  const data = await response.json();
  testLogger.info(`RUM token retrieved: ${data.token}`);

  return data.token;
}

/**
 * Generate expected stacktrace with actual content hashes
 * @param {string} errorType - Error type (e.g., 'typeError')
 * @param {Object} hashes - Content hashes from build
 * @returns {string} Stacktrace with actual hashes
 */
function getExpectedStacktrace(errorType, hashes) {
  const templates = {
    typeError: `TypeError: Cannot read properties of undefined (reading 'profile')
    at Ht @ http://localhost:8089/main.${hashes.main}.js:1:186484
    at http://localhost:8089/main.${hashes.main}.js:1:186818`,

    referenceError: `ReferenceError: undeclaredVariableThatDoesNotExist is not defined
    at Zt @ http://localhost:8089/main.${hashes.main}.js:1:186511`,

    rangeError: `RangeError: Invalid array length
    at Kt @ http://localhost:8089/main.${hashes.main}.js:1:186576`,

    customError: `PaymentError: Insufficient funds in account
    at Jt @ http://localhost:8089/main.${hashes.main}.js:1:186730`,

    crossChunk: `TypeError: Cannot read properties of null (reading 'timestamp')
    at u @ http://localhost:8089/lazy-module.${hashes.lazyModule}.js:1:131
    at http://localhost:8089/main.${hashes.main}.js:1:186963`
  };

  if (!templates[errorType]) {
    throw new Error(`Unknown error type: ${errorType}`);
  }

  return templates[errorType];
}

/**
 * Trigger errors in test app via Playwright
 * @param {Object} page - Playwright page
 * @param {string} testAppUrl - Test app URL
 * @returns {Promise<void>}
 */
async function triggerErrors(page, testAppUrl) {
  testLogger.info(`Navigating to test app: ${testAppUrl}`);

  await page.goto(testAppUrl);
  await page.waitForLoadState('networkidle', { timeout: 10000 });

  // Wait for SDK to initialize
  await page.waitForTimeout(2000);

  // Click all error buttons
  const errorButtons = [
    '#btn-type-error',
    '#btn-ref-error',
    '#btn-range-error',
    '#btn-custom-error',
    '#btn-nested-error'
    // Note: promise-error, async-error, and lazy-error are caught by .catch() handlers
  ];

  for (const buttonId of errorButtons) {
    testLogger.info(`Clicking ${buttonId}...`);
    await page.locator(buttonId).click();
    await page.waitForTimeout(500); // Small delay between clicks
  }

  testLogger.info('All errors triggered');

  // Wait for SDK to flush (30 seconds)
  testLogger.info('Waiting 30 seconds for SDK to flush events...');
  await page.waitForTimeout(30000);

  testLogger.info('Error ingestion complete');
}

/**
 * Create test user for RBAC testing
 * @param {Object} request - Playwright APIRequestContext
 * @param {Object} config - Configuration
 * @param {string} role - User role (viewer, editor, admin)
 * @returns {Promise<Object>} User info
 */
async function createTestUser(request, config, role) {
  const { baseUrl, org, auth } = config;

  const email = `test-${role}-${Date.now()}@sourcemap-test.local`;
  const password = 'TestPass123!';

  testLogger.info(`Creating test user: ${email} with role: ${role}`);

  const response = await request.post(`${baseUrl}/api/${org}/users`, {
    headers: {
      'Authorization': `Basic ${auth}`,
      'Content-Type': 'application/json'
    },
    data: {
      email,
      password,
      first_name: 'Test',
      last_name: role.charAt(0).toUpperCase() + role.slice(1),
      role
    }
  });

  const status = response.status();

  if (status !== 200 && status !== 201) {
    const body = await response.text();
    testLogger.error(`Failed to create user: ${body}`);
    throw new Error(`Failed to create user: ${status}`);
  }

  testLogger.info(`User created successfully: ${email}`);

  return {
    email,
    password,
    role,
    auth: Buffer.from(`${email}:${password}`).toString('base64')
  };
}

/**
 * Delete test user
 * @param {Object} request - Playwright APIRequestContext
 * @param {Object} config - Configuration
 * @param {string} email - User email
 * @returns {Promise<void>}
 */
async function deleteTestUser(request, config, email) {
  const { baseUrl, org, auth } = config;

  testLogger.info(`Deleting test user: ${email}`);

  const response = await request.delete(
    `${baseUrl}/api/${org}/users/${encodeURIComponent(email)}`,
    {
      headers: { 'Authorization': `Basic ${auth}` }
    }
  );

  const status = response.status();

  if (status !== 200) {
    const body = await response.text();
    testLogger.warn(`Delete user returned ${status}: ${body}`);
  } else {
    testLogger.info(`User deleted: ${email}`);
  }
}

module.exports = {
  buildTestApp,
  serveTestApp,
  stopTestAppServer,
  uploadSourcemaps,
  listSourcemaps,
  deleteSourcemaps,
  resolveStacktrace,
  getRumToken,
  getExpectedStacktrace,
  triggerErrors,
  createTestUser,
  deleteTestUser
};

File: tests/api-testing/helpers/sourcemap_helpers.py
"""
Sourcemap Test Helpers for Pytest

Utility functions for sourcemap API testing.
"""

import os
import subprocess
import time
import signal
import logging
import json
from pathlib import Path

logger = logging.getLogger(__name__)

# Path to static fixtures (pre-built sourcemaps)
FIXTURES_DIR = Path(__file__).parent.parent / "fixtures" / "sourcemaps"

# Path to test app (deprecated - tests now use static fixtures)
TEST_APP_DIR = Path(__file__).parent.parent.parent / "ui-testing" / "MD_Files" / "Sourcemaps" / "o2-sourcemap-test-app"


def load_static_sourcemaps(config=None):
    """
    Load pre-built sourcemaps from fixtures directory.

    This replaces the dynamic build approach - tests use static sourcemaps
    committed to the repository instead of building the test app.

    Args:
        config (dict): Configuration options (service, env, version, org)

    Returns:
        dict: Build result compatible with test expectations
    """
    # Default config
    if config is None:
        config = {}

    service = config.get('service', 'o2-sourcemap-test-app')
    env = config.get('env', 'testing')
    version = config.get('version', '1.0.0-static')
    org = config.get('org', 'default')

    logger.info(f"Loading static sourcemaps: service={service}, env={env}, version={version}")

    # Load hashes from JSON
    hashes_file = FIXTURES_DIR / "hashes.json"
    if not hashes_file.exists():
        raise FileNotFoundError(f"Hashes file not found: {hashes_file}")

    with open(hashes_file, 'r') as f:
        hashes = json.load(f)

    logger.info(f"Loaded content hashes: {hashes}")

    # Path to sourcemaps ZIP
    sourcemaps_zip = FIXTURES_DIR / "sourcemaps.zip"
    if not sourcemaps_zip.exists():
        raise FileNotFoundError(f"Sourcemaps ZIP not found: {sourcemaps_zip}")

    logger.info(f"Using sourcemaps ZIP: {sourcemaps_zip}")

    return {
        'app_dir': str(FIXTURES_DIR),
        'dist_path': str(FIXTURES_DIR),
        'hashes': hashes,
        'sourcemaps_zip': str(sourcemaps_zip),
        'config': {'service': service, 'env': env, 'version': version, 'org': org}
    }


def build_test_app(rum_token, config=None):
    """
    Build the sourcemap test app with updated RUM SDK configuration.

    Args:
        rum_token (str): RUM token from API
        config (dict): Build configuration options

    Returns:
        dict: Build result with hashes and paths
    """
    if config is None:
        config = {}

    site = config.get('site', 'localhost:5080')
    insecure_http = config.get('insecureHTTP', True)
    service = config.get('service', 'o2-sourcemap-test-app')
    env = config.get('env', 'testing')
    version = config.get('version', '1.0.0-api-test')
    org = config.get('org', 'default')

    logger.info(f"Building test app: service={service}, env={env}, version={version}")

    # Path to config file
    config_path = TEST_APP_DIR / "src" / "index.js"

    # Read current config
    with open(config_path, 'r') as f:
        config_content = f.read()

    # Update SDK config
    import re
    config_content = re.sub(r"clientToken:\s*['\"][^'\"]*['\"]", f"clientToken: '{rum_token}'", config_content)
    config_content = re.sub(r"site:\s*['\"][^'\"]*['\"]", f"site: '{site}'", config_content)
    config_content = re.sub(r"insecureHTTP:\s*(?:true|false)", f"insecureHTTP: {str(insecure_http).lower()}", config_content)
    config_content = re.sub(r"service:\s*['\"][^'\"]*['\"]", f"service: '{service}'", config_content)
    config_content = re.sub(r"env:\s*['\"][^'\"]*['\"]", f"env: '{env}'", config_content)
    config_content = re.sub(r"version:\s*['\"][^'\"]*['\"]", f"version: '{version}'", config_content)
    config_content = re.sub(r"organizationIdentifier:\s*['\"][^'\"]*['\"]", f"organizationIdentifier: '{org}'", config_content)

    # Write updated config
    with open(config_path, 'w') as f:
        f.write(config_content)

    logger.info("Updated SDK configuration")

    # Build app
    logger.info("Running npm run build...")
    result = subprocess.run(
        ['npm', 'run', 'build'],
        cwd=TEST_APP_DIR,
        capture_output=True,
        text=True,
        timeout=60
    )

    if result.returncode != 0:
        logger.error(f"Build failed: {result.stderr}")
        raise Exception(f"Build failed: {result.stderr}")

    logger.info("Build completed successfully")

    # Extract content hashes
    dist_path = TEST_APP_DIR / "dist"
    dist_files = os.listdir(dist_path)

    hashes = {
        'main': extract_hash(dist_files, 'main', '.js'),
        'lazy_module': extract_hash(dist_files, 'lazy-module', '.js'),
        'profiler': extract_hash(dist_files, 'profiler', '.js'),
        'recorder': extract_hash(dist_files, 'recorder', '.js')
    }

    logger.info(f"Extracted hashes: {hashes}")

    # Create sourcemaps ZIP
    sourcemaps_zip = dist_path / "sourcemaps.zip"

    # Remove old ZIP if exists
    if sourcemaps_zip.exists():
        sourcemaps_zip.unlink()

    subprocess.run(
        ['zip', 'sourcemaps.zip'] + [f for f in dist_files if f.endswith('.map')],
        cwd=dist_path,
        check=True,
        capture_output=True
    )

    logger.info(f"Created sourcemaps ZIP at {sourcemaps_zip}")

    return {
        'app_dir': str(TEST_APP_DIR),
        'dist_path': str(dist_path),
        'hashes': hashes,
        'sourcemaps_zip': str(sourcemaps_zip),
        'config': {'service': service, 'env': env, 'version': version, 'org': org}
    }


def extract_hash(files, prefix, ext):
    """Extract content hash from filename."""
    import re
    pattern = re.compile(f"{prefix}\\.(.*?)\\{ext}")

    for file in files:
        if file.startswith(prefix) and file.endswith(ext) and not file.endswith('.map'):
            match = pattern.search(file)
            if match:
                return match.group(1)

    raise Exception(f"Could not find file matching {prefix}*{ext}")


def serve_test_app(dist_path, port=8089):
    """
    Start HTTP server to serve test app.

    Args:
        dist_path (str): Path to dist directory
        port (int): Port number

    Returns:
        dict: Server process info
    """
    logger.info(f"Starting HTTP server on port {port}...")

    # Start Python HTTP server
    process = subprocess.Popen(
        ['python3', '-m', 'http.server', str(port)],
        cwd=dist_path,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        preexec_fn=os.setsid  # Create new process group
    )

    # Wait for server to be ready
    time.sleep(2)

    logger.info(f"HTTP server started (PID: {process.pid})")

    return {
        'process': process,
        'pid': process.pid,
        'port': port,
        'url': f'http://localhost:{port}'
    }


def stop_test_app_server(server_info):
    """Stop the HTTP server."""
    if not server_info or 'pid' not in server_info:
        logger.warning("No server PID provided, skipping stop")
        return

    pid = server_info['pid']
    logger.info(f"Stopping HTTP server (PID: {pid})...")

    try:
        # Kill process group
        os.killpg(os.getpgid(pid), signal.SIGTERM)
        logger.info("HTTP server stopped")
    except Exception as e:
        logger.warning(f"Failed to stop server: {e}")
        try:
            # Try direct kill
            os.kill(pid, signal.SIGKILL)
        except:
            pass

    # Also kill any processes on the port
    port = server_info.get('port', 8089)
    try:
        # Use list-based args to avoid shell injection
        result = subprocess.run(
            ['lsof', '-ti', f':{port}'],
            capture_output=True,
            text=True,
            check=False
        )
        if result.stdout.strip():
            pids = result.stdout.strip().split('\n')
            for pid in pids:
                try:
                    os.kill(int(pid), signal.SIGKILL)
                except (ValueError, ProcessLookupError):
                    pass
    except Exception:
        pass


def upload_sourcemaps(session, base_url, org_id, zip_path, service, env, version):
    """Upload sourcemaps to OpenObserve."""
    logger.info(f"Uploading sourcemaps: service={service}, env={env}, version={version}")

    with open(zip_path, 'rb') as f:
        files = {'file': ('sourcemaps.zip', f, 'application/zip')}
        data = {
            'service': service,
            'env': env,
            'version': version
        }

        response = session.post(
            f"{base_url}api/{org_id}/sourcemaps",
            files=files,
            data=data
        )

    logger.info(f"Upload response: {response.status_code}")

    if response.status_code not in [200, 201]:
        logger.error(f"Upload failed: {response.text}")
        raise Exception(f"Upload failed: {response.status_code} - {response.text}")

    return response


def list_sourcemaps(session, base_url, org_id, service=None, env=None, version=None):
    """List sourcemaps from OpenObserve."""
    params = {}
    if service:
        params['service'] = service
    if env:
        params['env'] = env
    if version:
        params['version'] = version

    response = session.get(
        f"{base_url}api/{org_id}/sourcemaps",
        params=params
    )

    if response.status_code != 200:
        logger.error(f"List failed: {response.text}")
        raise Exception(f"List failed: {response.status_code}")

    return response.json()


def delete_sourcemaps(session, base_url, org_id, service, env, version):
    """Delete sourcemaps from OpenObserve."""
    logger.info(f"Deleting sourcemaps: service={service}, env={env}, version={version}")

    params = {
        'service': service,
        'env': env,
        'version': version
    }

    response = session.delete(
        f"{base_url}api/{org_id}/sourcemaps",
        params=params
    )

    logger.info(f"Delete response: {response.status_code}")

    return response


def resolve_stacktrace(session, base_url, org_id, service, env, version, stacktrace):
    """Resolve minified stacktrace via sourcemap API."""
    logger.info(f"Resolving stacktrace for {service}:{version}")

    payload = {
        'service': service,
        'env': env,
        'version': version,
        'stacktrace': stacktrace
    }

    response = session.post(
        f"{base_url}api/{org_id}/sourcemaps/stacktrace",
        json=payload
    )

    if response.status_code != 200:
        logger.error(f"Resolution failed: {response.text}")
        raise Exception(f"Resolution failed: {response.status_code}")

    return response.json()


def get_expected_stacktrace(error_type, hashes):
    """Generate expected stacktrace with actual content hashes."""
    templates = {
        'type_error': f"""TypeError: Cannot read properties of undefined (reading 'profile')
    at Ht @ http://localhost:8089/main.{hashes['main']}.js:1:186484
    at http://localhost:8089/main.{hashes['main']}.js:1:186818""",

        'reference_error': f"""ReferenceError: undeclaredVariableThatDoesNotExist is not defined
    at Zt @ http://localhost:8089/main.{hashes['main']}.js:1:186511""",

        'range_error': f"""RangeError: Invalid array length
    at Kt @ http://localhost:8089/main.{hashes['main']}.js:1:186576""",

        'custom_error': f"""PaymentError: Insufficient funds in account
    at Jt @ http://localhost:8089/main.{hashes['main']}.js:1:186730""",

        'cross_chunk': f"""TypeError: Cannot read properties of null (reading 'timestamp')
    at u @ http://localhost:8089/lazy-module.{hashes['lazy_module']}.js:1:131
    at http://localhost:8089/main.{hashes['main']}.js:1:186963"""
    }

    if error_type not in templates:
        raise ValueError(f"Unknown error type: {error_type}")

    return templates[error_type]


def get_rum_token(session, base_url, org_id):
    """Get RUM token from OpenObserve."""
    logger.info("Fetching RUM token...")

    response = session.get(f"{base_url}api/{org_id}/rumtoken")

    if response.status_code != 200:
        logger.error(f"Failed to get RUM token: {response.text}")
        raise Exception(f"Failed to get RUM token: {response.status_code}")

    data = response.json()
    # Use same pattern as ingest_rum_errors.py to handle nested response structure
    token = data.get('data', {}).get('rum_token') or data.get('rum_token')

    if not token:
        logger.error(f"RUM token is empty in response: {data}")
        raise Exception("RUM token not found in response")

    logger.info(f"RUM token retrieved: {token}")

    return token


# Expected resolution results for validation
EXPECTED_RESOLUTIONS = {
    'type_error': {
        'file': 'src/errors/sync-errors.js',
        'line': 10,
        'column': 15,
        'function': 'profile',
        'contains': 'return user.profile.name'
    },
    'reference_error': {
        'file': 'src/errors/sync-errors.js',
        'line': 16,
        'contains': 'undeclaredVariableThatDoesNotExist'
    },
    'range_error': {
        'file': 'src/errors/sync-errors.js',
        'line': 22,
        'contains': 'new Array(-1)'
    },
    'custom_error': {
        'file': 'src/errors/sync-errors.js',
        'line': 36,
        'contains': 'PaymentError'
    },
    'cross_chunk': {
        'file': 'src/errors/lazy-module.js',
        'line': 20,
        'function': 'triggerLazyModuleError'
    }
}
